Meta: Add an experimental GN build

ADKaster opened the pull request:
This PR adds the ability to build ladybird with the GN build system

https://gn.googlesource.com/gn/#gn

The build is incomplete in that it only knows how to build Ladybird, and not serenity itself. GN offers more ways to parallelize the build than CMake, while supporting cross-compilation and "host tools" in a more natural and native way than CMake.

On my MacBook Pro, building ladybird from 0 ccache and zero build directory with the CMake build takes this long:

```
[2780/2780] Linking CXX executable Ladybird/ladybird.app/Contents/MacOS/ladybird

real	3m7.292s
user	21m38.676s
sys	2m50.478s
```

While using gn it takes this long:

```
$ gn gen out && time ninja -C out
Done. Made 2603 targets from 130 files in 142ms
ninja: Entering directory `out'
[23/5930] ACTION //Userland/Libraries/LibTLS:ca_certificates_download(//Meta/gn/build/toolchain:unix)
Downloading version 2023-01-10 of cacert.pem...done
[46/5930] ACTION //Userland/Libraries/LibTimeZone:timezone_database_download(//Meta/gn/build/toolchain:unix)
Downloading version 2023b of gen/Cache/TZDB/tzdb.tar.gz...done
[3044/5930] ACTION //Userland/Libraries/LibUnicode:emoji_test_download(//Meta/gn/build/toolchain:unix)
Downloading version 15.0 of gen/Cache/EMOJI/emoji-test.txt...done
[3460/5930] ACTION //Userland/Libraries/LibUnicode:unicode_database_download(//Meta/gn/build/toolchain:unix)
Downloading version 15.0.0 of gen/Cache/UCD/UCD.zip...done
[3682/5930] ACTION //Userland/Libraries/LibLocale:locale_database_download(//Meta/gn/build/toolchain:unix)
Downloading version 42.0.0 of gen/Cache/CLDR/cldr.zip...done
[5930/5930] STAMP obj/default.stamp

real	1m53.155s
user	17m54.816s
sys	1m34.867s
```

Those numbers are kind of cheating since the CMake build does all its file downloads in the configure step. If we blast Build/lagom but keep Build/Caches and run again without ccache, the CMake build takes only

```
real	2m22.117s
user	20m29.872s
sys	2m34.071s
```

cc @nico @trflynn89

ADKaster commented on Meta/gn/secondary/Userland/Libraries/LibTest/BUILD.gn:
```diff
@@ -0,0 +1,52 @@
+  testonly = true
+}
+
+# FIXME: These are OBJECT libraries in the CMake. Is source_set() the correct choice?
```
Does gn deduplicate source_sets with the same flags? i.e. is `source_set` the right translation from object library here? or should they be static libs? the fact that it's got main() in it makes me a bit cautious.

nico replied:
For a source_set, gn compiles all the source files in it to .o files and then adds all the .o files in it to the first linkable thing that (transitively) depends on it. I think that's what OBJECT libraries do, too. I translate OBJECT libs to source_sets.

But I don't understand your "deduplicate" question.

(In particular, this means all .o files in a source_set are always linked in. A static library on the other hand produce a .a file, which has a symbol->.o file map at the start, and only the .o files in it that are needed by unresolved symbols at link time are pulled in by the linker.)

ADKaster replied:
The de-duplicate question is more about thinking that maybe a `source_set` would simply add the files to `sources` of every target that depends on them and then compile the sources once for each each target. Given your explanation above, it sounds like they are smarter than that and do behave like source_sets. so I'll remove that fixme question

trflynn89 commented:
A couple of flyby comments:
* It'd be a lot more ergonomic if the BUILD.gn were in the the folder of the sources themselves, rather than having them in //Meta/gn/secondary/... (or is this just temporary while the GN build is experimental?)
* Let's use 4 spaces for tabs instead of google's 2-space style? :) (at least for files not borrowed from LLVM)
* Do we need to attribute LLVM or Google for those files we grabbed from them?

BertalanD commented on Meta/gn/build/buildflags.gni:
```diff
@@ -0,0 +1,18 @@
+  # Whether to build with optimizations.
+  is_optimized = !is_debug
+
+  if (is_debug) {
+    # Debug info symbol level. 0: No symbols; 1: Line numbers; 2: Full symbols.
+    symbol_level = 2
+  } else {
+    # Debug info symbol level. 0: No symbols; 1: Line numbers; 2: Full symbols.
+    symbol_level = 0
+  }
```
Does this mean that GN-built Ladybird does not have any debug info?

ADKaster replied:
Yes, unless you manipulate the gn args (such as `gn args out --args=symbol_level=1`) to get the symbol level to 1  or higher.

BertalanD replied:
Could that be accountable for some of the build speed increase? IIRC our CMake Lagom build unconditionally includes line tables (`-g1`).

ADKaster replied:
Some? Definitely not all of it. With `symbol_level=1` out/args.gn, I got this:

```
real	1m54.314s
user	18m19.242s
sys	1m40.234s
```

I'm personally convinced that the build speedup is organizing the jobs better, and doing less work around generated code. It's possible the missing c/cxxflags are causing the compiler to do less work for each cpp file though 🤷

ADKaster replied:
I've got local changes to default to symbol_level=1

ADKaster commented:
> It'd be a lot more ergonomic if the BUILD.gn were in the the folder of the sources themselves, rather than having them in //Meta/gn/secondary/... (or is this just temporary while the GN build is experimental?)

Intended to just be an experimental thing. If we want to adopt it as a more official secondary build system then I can for sure get rid of the secondary build directory and put the files in the standard tree.

> Let's use 4 spaces for tabs instead of google's 2-space style? :) (at least for files not borrowed from LLVM)

I'm not sure if we can convince ``gn format``  to give us four spaces, it's pretty opinionated on its style. If we want to fully manually format the files then we can go with two spaces though 🤷 

> Do we need to attribute LLVM or Google for those files we grabbed from them?

I'm not sure honestly. The LLVM files are quite similar to ones in Chromium (since Nico added them). They don't have a license identifier in LLVM, but their counterparts in chromium seem to. The ones I would think are copy-paste enough would be the BUILDCONFIG.gn and its friends (the gnu files for that), and the write_cmake_config.py/gni, and compiled_action.gni The rest of the lifted files are either trivial or substantially modified imo.

nico commented on Meta/gn/secondary/Meta/Lagom/Tools/CodeGenerators/LibLocale/BUILD.gn:
```diff
@@ -0,0 +1,42 @@
+  deps = [
+    "//Userland/Libraries/LibMain",
+    "//Userland/Libraries/LibTimeZone",
+    "//Meta/Lagom/Tools/CodeGenerators/LibUnicode:headers",
```
This looks kind of weird.

Usually .h and corresponding .cpp files are in the same target, and then if you dep on that target, you link in the definitions for the declarations in the .h files. A source_set() with just .h files has no link-time effect at all (headers aren't compiled into anything by themselves). I suppose this would make sense for a .h file that contains only inline functions. Is that the case here?

ADKaster replied:
This was an attempt to fix a GN check error from this construct, where other code generators in other libs include a header from LibUnicode's code generator:

https://github.com/SerenityOS/serenity/blob/36ff6187f6f5e67ba3f2085b5257ec1de10cf79b/Meta/Lagom/Tools/CodeGenerators/LibLocale/GenerateDateTimeFormatData.cpp#L7

I believe GeneratorUtil.h is entirely header only, yes. It's a lot of templates.

ADKaster replied:
I'll remove this silly header only source set until we fix the fixme in code.

nico commented on Meta/gn/secondary/Meta/Lagom/Tools/CodeGenerators/LibUnicode/BUILD.gn:
```diff
@@ -0,0 +1,21 @@
+import("//Meta/Lagom/Tools/lagom_tool.gni")
+
+source_set("headers") {
+    sources = [ "GeneratorUtil.h" ]
```
run `git ls-files '*.gn' '*.gni' | xargs gn format`

ADKaster replied:
formatted the whole lot.

nico commented on Meta/gn/secondary/Meta/Lagom/Tools/lagom_tool.gni:
```diff
@@ -0,0 +1,16 @@
+template("lagom_tool") {
+  if (current_toolchain == host_toolchain) {
```
This should only be used in lagom files, right? Maybe `assert(current_toolchain == host_toolchain)` instead?

ADKaster replied:
One day we might want to instal these tools (cross-compiled) onto the system so that the self-hosting story is a bit easier, but for now yeah, that's true. I'll assert host_toolchain.

nico commented on Meta/gn/secondary/Userland/Libraries/LibTimeZone/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+import("//Meta/gn/build/library_name.gni")
+
+declare_args() {
+    enable_timezone_database_download = false
```
If you add a one-line comment above this, it'll be printed as help on `gn args --list out/gn`

nico commented on Meta/gn/build/download_file.py:
```diff
@@ -0,0 +1,60 @@
+    print(f"Downloading version {args.version} of {args.output}...", end='')
+
+    with urllib.request.urlopen(args.url) as f:
+        with open(args.output, 'wb') as out:
```
worth downloading to a temp location and then renaming when the download is done, so that an interrupted download doesn't lead to an incorrect incremental build?

ADKaster replied:
sure, using a temp file here now (though a bit ugly, there's probably a cleaner way)

nico commented on Meta/gn/build/extract_archive_contents.py:
```diff
@@ -0,0 +1,76 @@
+            except KeyError:
+                pass
+            # Assume that this is actually a *directory* rather than a filename
+            # FIXME: Do this smarter... Can we convince GN to let us put directorie
```
You can write as many files as you want without telling the build system about it. You just can't have build system dependency chains for those files.

ADKaster replied:
Hm sounds like in this case the "answer" is to pass the args differently for the CLDR invocation that wants to extract (8k!) files into ~8 directories rather than N < 10 individual files.

nico commented on Meta/gn/secondary/Userland/Libraries/LibTimeZone/BUILD.gn:
```diff
@@ -32,16 +33,37 @@ if (enable_timezone_database_download) {
   include_dirs = [
     "//Userland/Libraries",
+    # FIXME: How to get the parent of the current target's gen dir?
+    "$target_gen_dir/..",
```
What do you need this for?

I sometimes use a public_config to push an include dir to deps: https://github.com/llvm/llvm-project/blob/main/llvm/utils/gn/secondary/lld/Common/BUILD.gn#L4 (but sometimes I'm lazy and just add something to include_dirs)

ADKaster replied:
A common pattern for generated includes like this one is to #include them like so: 

```
#include <LibTimeZone/TimeZoneData.h>
```

Which assumes that we have the *parent* of the current lib's binary directory (or generated file directory) passes as an include dir. It's a bit... awkward of a pattern. I'm probably overcomplicating it though. The thing that the CMake build does is add `${CMAKE_BINARY_DIR}/Userland/Libraries` and `${CMAKE_BINARY_DIR}/Userland/Services/` to everyone's path with `include_directories()` at the top level.

ADKaster replied:
Since this is the expected pattern, I'll just remove the fixme. It could definitely be done smarter though.

nico commented on Meta/gn/secondary/Userland/Libraries/LibGL/BUILD.gn:
```diff
@@ -0,0 +1,57 @@
+    "-c", rebase_path(outputs[1], root_build_dir),
+    "-j", rebase_path(inputs[0], root_build_dir),
+  ]
+  # FIXME: install header into $prefix/include/LibGL/GL on serenity
```
here's a header "install" example: https://github.com/llvm/llvm-project/blob/main/llvm/utils/gn/secondary/libcxx/include/BUILD.gn#L75

nico commented on Meta/gn/secondary/Userland/Libraries/LibWebView/BUILD.gn:
```diff
@@ -0,0 +1,85 @@
+  tool = "//Meta/Lagom/Tools/CodeGenerators/IPCCompiler"
+  inputs = [ "//Userland/Services/WebContent/WebContentServer.ipc" ]
+  outputs = [
+    "$root_gen_dir/WebContent/WebContentServerEndpoint.h",
```
(run `gn format`)

nico reviewed (commented):
Got until "Meta: Add gn build rules for LibWeb". Maybe some of these comments are useful?

ADKaster commented on Meta/gn/build/toolchain/BUILD.gn:
```diff
@@ -0,0 +1,141 @@
+unix_toolchain("unix") {
+  toolchain_args = {
+    current_os = host_os
+    #current_cpu = host_cpu
```
It's not used in any of the current build rules, so it was complained to be unused 😅 . I can remove it if we don't need it, but I wasn't sure at the beginning

ADKaster commented on Meta/gn/build/library_name.gni:
```diff
@@ -0,0 +1,6 @@
+declare_args() {
+  output_name_prefix = ""
+  if (current_os != "serenity") {
+    output_name_prefix = "lagom-"
```
Good point. I've got local changes to set it in the `alink` and `solink` tools. Not sure what to do about `solink_module`, since the llvm build doesn't seem to set the output_name in that case.

ADKaster commented:
Updates since last reviewed:
- formatted all files with `gn format`
- default to symbol_level=1
- removed library_name.gni and set `liblagom-` prefix in toolchain/BUILD.gn
- documented args
- removed unnecessary FIXMEs
- couple other housekeeping things Nico noticed

BenWiederhake2 reviewed (commented):
Looks nice! I just did a high level flyover and found a typo.

ADKaster commented on Meta/gn/build/embed_as_string_view.gni:
```diff
@@ -0,0 +1,64 @@
+# This file introduces a templates for calling embed_as_string_view.py.
```
thanks! I fixed up the grammar in the other template files with this verbiage.

BenWiederhake replied:
Only now I see that I typo'd in a comment about fixing a typo. Shame on me, and thanks for not copying my own typo :D

I originally meant to suggest either "introduces templates" or "introduces a template".

ADKaster commented:
Updates since last: 
- Fixed up build so that it works on Ubuntu 22.04 with gn args as below

```
host_cc = "clang"
host_cxx = "clang++"
is_clang = true
use_lld = true
```
